cmd/internal/kube/daemonset/kube_events_worker.go

```diff
@@ -136,59 +163,102 @@ func (d *Daemonset) handlePodModifyEvent(pod coreV1.Pod) {
 // in the pod, fetches the environment variables of that container, and extracts the
 // necessary variables such as the project ID, API key, and environment.
 func (d *Daemonset) inspectPodForEnvVars(pod coreV1.Pod, podArgs *PodArgs) error {
-	// Get the UUID of the main container in the pod
-	containerUUID, err := d.KubeClient.GetMainContainerUUID(pod)
+	// Get the UUIDs of all containers in the pod
+	containerUUIDs, err := d.KubeClient.GetContainerUUIDs(pod)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get main container UUID for pod: %s", pod.Name)
+		return errors.Wrapf(err, "failed to get container UUIDs for pod: %s", pod.Name)
 	}
 
-	// Get the environment variables of the main container
-	envVars, err := d.CRIClient.GetEnvVars(containerUUID)
-	if err != nil {
-		return errors.Wrapf(err, "failed to get environment variables for pod/container : %s/%s", pod.Name, containerUUID)
+	if len(containerUUIDs) == 0 {
+		return errors.New("no running containers found in the pod")
 	}
 
-	var (
-		insightsProjectID akid.ServiceID
-		insightsAPIKey    string
-	)
+	containerConfigMap := maps.NewMap[string, requiredContainerConfig]()
 
-	// Extract the necessary environment variables
-	for key, value := range envVars {
-		switch key {
-		case POSTMAN_INSIGHTS_PROJECT_ID:
-			err := akid.ParseIDAs(value, &insightsProjectID)
-			if err != nil {
-				return errors.Wrap(err, "failed to parse project ID")
-			}
-		case POSTMAN_INSIGHTS_API_KEY:
-			insightsAPIKey = value
+	// Iterate over all containers in the pod to check for the required environment variables
+	for _, containerUUID := range containerUUIDs {
+		envVars, err := d.CRIClient.GetEnvVars(containerUUID)
+		if err != nil {
+			printer.Debugf("Failed to get environment variables for pod/container : %s/%s\n", pod.Name, containerUUID)
+			continue
 		}
+
+		containerEnvVars := requiredContainerConfig{}
+		if projectID, exists := envVars[POSTMAN_INSIGHTS_PROJECT_ID]; exists {
+			containerEnvVars.projectID = projectID
+		}
+		if apiKey, exists := envVars[POSTMAN_INSIGHTS_API_KEY]; exists {
+			containerEnvVars.apiKey = apiKey
+		}
+		containerConfigMap[containerUUID] = containerEnvVars
 	}
 
-	if (insightsProjectID == akid.ServiceID{}) && insightsAPIKey == "" {
-		return allRequiredEnvVarsAbsentErr
+	mainContainerUUID := ""
+	mainContainerConfig := requiredContainerConfig{}
+	mainContainerMissingAttrs := []string{}
+	maxSetAttrs := -1
+
+	for uuid, config := range containerConfigMap {
+		attrCount, missingAttrs := countSetAttributes(config)
+		if attrCount > maxSetAttrs {
+			maxSetAttrs = attrCount
+			mainContainerMissingAttrs = missingAttrs
+			mainContainerUUID = uuid
+			mainContainerConfig = config
+		}
```

Comment:
I think you could have used an array and then used `slices.MaxFunc` to extract the best option (or done this comparison inline)  -- we don't really need to look up by UUID.  But this is OK.